### ISSUE TITLE
Consider using semantic versioning scheme

### ISSUE THREAD

**Issue by sanjayankur31** [vxl/vxl#624]
Would it please be possible to version the shared objects that we can create from the vxl libraries? At the moment, all the so files are unversioned. This makes it harder for people to track API/ABI changes.

```
$ ls
libbzip2.a     libnetlib.a            libtestlib.a                 libvbl_io.a    libvgl_algo.so  libvil_algo.a   libvnl_io.so     libvpgl_file_formats.a   libvsl.a
libbzip2.so    libnetlib.so           libtestlib.so                libvbl_io.so   libvgl_io.a     libvil_algo.so  libvnl.so        libvpgl_file_formats.so  libvsl.so
libclipper.a   libopenjpeg2.a         libtiff.a                    libvbl.so      libvgl_io.so    libvil_io.a     libvnl_xio.a     libvpgl_io.a             libvul.a
libclipper.so  libopenjpeg2.so        libtiff.so                   libvcl.a       libvgl.so       libvil_io.so    libvnl_xio.so    libvpgl_io.so            libvul_io.a
libdcmtk.a     libopenjpeg2.so.2.0.0  libv3p_netlib.a              libvcl.so      libvgl_xio.a    libvil.so       libvpdl.a        libvpgl.so               libvul_io.so
libdcmtk.so    libpng.a               libv3p_netlib.so             libvcsl.a      libvgl_xio.so   libvnl.a        libvpdl.so       libvpgl_xio.a            libvul.so
libgeotiff.a   libpng.so              libvbl.a                     libvcsl.so     libvil1.a       libvnl_algo.a   libvpgl.a        libvpgl_xio.so           libz.a
libjpeg.a      librply.a              libvbl_example_templates.a   libvgl.a       libvil1.so      libvnl_algo.so  libvpgl_algo.a   libvpl.a                 libz.so
libjpeg.so     librply.so             libvbl_example_templates.so  libvgl_algo.a  libvil.a        libvnl_io.a     libvpgl_algo.so  libvpl.so
(ins)[asinha@ankur-pc  lib]$ for f in *.so; do readelf -d $f | grep SONAME; done
 0x000000000000000e (SONAME)             Library soname: [libbzip2.so]
 0x000000000000000e (SONAME)             Library soname: [libclipper.so]
 0x000000000000000e (SONAME)             Library soname: [libdcmtk.so]
 0x000000000000000e (SONAME)             Library soname: [libjpeg.so]
 0x000000000000000e (SONAME)             Library soname: [libnetlib.so]
 0x000000000000000e (SONAME)             Library soname: [libopenjpeg2.so.2.0.0]
 0x000000000000000e (SONAME)             Library soname: [libpng.so]
 0x000000000000000e (SONAME)             Library soname: [librply.so]
 0x000000000000000e (SONAME)             Library soname: [libtestlib.so]
 0x000000000000000e (SONAME)             Library soname: [libtiff.so]
 0x000000000000000e (SONAME)             Library soname: [libv3p_netlib.so]
 0x000000000000000e (SONAME)             Library soname: [libvbl_example_templates.so]
 0x000000000000000e (SONAME)             Library soname: [libvbl_io.so]
 0x000000000000000e (SONAME)             Library soname: [libvbl.so]
 0x000000000000000e (SONAME)             Library soname: [libvcl.so]
 0x000000000000000e (SONAME)             Library soname: [libvcsl.so]
 0x000000000000000e (SONAME)             Library soname: [libvgl_algo.so]
 0x000000000000000e (SONAME)             Library soname: [libvgl_io.so]
 0x000000000000000e (SONAME)             Library soname: [libvgl.so]
 0x000000000000000e (SONAME)             Library soname: [libvgl_xio.so]
 0x000000000000000e (SONAME)             Library soname: [libvil1.so]
 0x000000000000000e (SONAME)             Library soname: [libvil_algo.so]
 0x000000000000000e (SONAME)             Library soname: [libvil_io.so]
 0x000000000000000e (SONAME)             Library soname: [libvil.so]
 0x000000000000000e (SONAME)             Library soname: [libvnl_algo.so]
 0x000000000000000e (SONAME)             Library soname: [libvnl_io.so]
 0x000000000000000e (SONAME)             Library soname: [libvnl.so]
 0x000000000000000e (SONAME)             Library soname: [libvnl_xio.so]
 0x000000000000000e (SONAME)             Library soname: [libvpdl.so]
 0x000000000000000e (SONAME)             Library soname: [libvpgl_algo.so]
 0x000000000000000e (SONAME)             Library soname: [libvpgl_file_formats.so]
 0x000000000000000e (SONAME)             Library soname: [libvpgl_io.so]
 0x000000000000000e (SONAME)             Library soname: [libvpgl.so]
 0x000000000000000e (SONAME)             Library soname: [libvpgl_xio.so]
 0x000000000000000e (SONAME)             Library soname: [libvpl.so]
 0x000000000000000e (SONAME)             Library soname: [libvsl.so]
 0x000000000000000e (SONAME)             Library soname: [libvul_io.so]
 0x000000000000000e (SONAME)             Library soname: [libvul.so]
 0x000000000000000e (SONAME)             Library soname: [libz.so]
```

In the past, the Fedora maintainer has simply used the vxl release version as the soname version, but the API version and the software version do not necessarily need to match. I'd be much happier if upstream devs did this so that all of us downstream could follow the versioning that you set :)

cmake allows us to set the API soname version using the `SOVERSION` variable:

https://cmake.org/cmake/help/latest/prop_tgt/SOVERSION.html?highlight=soversion
https://stackoverflow.com/a/45843676/375067

**Comment by sanjayankur31** [comment]
@bradking  @hjmjohnson @andyneff @decrispell : could we stop the nightly timestamp bot now that we've begun versioning the shared objects? (The nightly commits make it exceptionally hard to follow code changes)

**Comment by decrispell** [comment]
I'm in favor of removing the nightly timestamp, with the caveat that I don't think it's realistic to expect the semantic version numbers to updated every time an API in contrib changes.  Is it reasonable to enforce the version bumps on core only?

**Comment by sanjayankur31** [comment]
I'm not well versed enough with the development model here. Don't the contrib maintainers also have to maintain API/ABI stability or inform users when it changes?

**Comment by decrispell** [comment]
At least for contrib/brl, the contents tend to be pretty specific research code, meaning the "developers" and "users" are usually the same people.   If a chunk of code turns out to be more generally useful, it can be promoted to core.  In cases where vxl is being used by another piece of software, we typically use git submodules to ensure compatibility.

**Comment by bradking** [comment]
> could we stop the nightly timestamp bot

I've turned off the bot.

**Comment by hjmjohnson** [comment]
@bradking Thank you.
@decrispell I agree,  I will make a GitHub PR template that reminds users to increment the API when the core API changes.  Would you be able to provide a good description of when that needs to be done?

**Comment by sanjayankur31** [comment]
> At least for contrib/brl, the contents tend to be pretty specific research code, meaning the "developers" and "users" are usually the same people. If a chunk of code turns out to be more generally useful, it can be promoted to core. In cases where vxl is being used by another piece of software, we typically use git submodules to ensure compatibility.

OK---that makes sense.

Could this be documented somewhere though? Maybe a README.md in the contrib folders that are meant to be experimental, and not stable?

**Comment by hjmjohnson** [comment]
I think all the issues have been addressed.
